qa/rpc-tests/feelimit.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Dogecoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Fee limit QA test.
+# Tests nodes under fee limit
+"""
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import *
+from decimal import Decimal
+
+class FeeLimitTest(BitcoinTestFramework):
+
+    def __init__(self):
+        super().__init__()
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+
+    def setup_network(self, split=False):
+        self.nodes = []
+        self.nodes.append(start_node(0, self.options.tmpdir, []))
+        self.nodes.append(start_node(1, self.options.tmpdir, []))
+        self.nodes.append(start_node(2, self.options.tmpdir, []))
+
+        connect_nodes_bi(self.nodes,0,1)
+        connect_nodes_bi(self.nodes,1,2)
+        connect_nodes_bi(self.nodes,0,2)
+
+        self.is_network_split=False
+        self.sync_all()
+
+    def run_test(self):
+        self.fee = Decimal("0.001")
+        self.seed = 1000
+
+        # mine some blocks and prepare some coins
+        self.nodes[2].generate(1)
+        self.sync_all()
+        self.nodes[0].generate(101)
+        self.sync_all()
+
+        # Check generating a default transaction is unchanged
+        txId = self.nodes[0].sendtoaddress(self.nodes[1].getnewaddress(), 10)
+        txObj = self.nodes[0].gettransaction(txId)
+        assert_equal(txObj['fee'], Decimal('-1.00000000'))
+        tx_size_kb = Decimal(count_bytes(txObj['hex'])) / Decimal("1000")
+        lower_fee_per_kb = Decimal("0.001")
+
+        # Force generating a transaction with 1.14.5-like fees by manually building the tx
+        fee = lower_fee_per_kb * tx_size_kb
+        utx = self.nodes[0].listunspent()[0]
+        inputs  = [ {'txid' : utx['txid'], 'vout' : utx['vout']}]
+        outputs = { self.nodes[2].getnewaddress() : utx['amount'] - fee }
+        rawtx   = self.nodes[0].createrawtransaction(inputs, outputs)
+        signedtx = self.nodes[0].signrawtransaction(rawtx)
+        singletxid = self.nodes[0].sendrawtransaction(signedtx["hex"])
```

Comment:
Needs a check before sending, making sure that the size of this signed tx is not one byte bigger than the previous tx (possible through low-then-high-R value in the scriptsig element across those tx as this has a 1-byte variance) or else this test will fail occasionally.

Alternatively, as a shortcut, add 1 byte to the size calculated on line 47 (that still tests the lack of rounding, yet allows for tests to not fail due to scriptsig size variance.)

Reply:
Am testing the proposed solution (currently not sure if re-signing the tx would trigger the same variance)

Reply:
My understanding is the signatures are now a stable length, so the risk would be if the value output changed significantly enough to alter length of the transaction output (the inputs of course do not change). That's only a concern with an order of magnitude change, though.

Do let me know if you spot something I missed, though.